File: framework/languages/java/e2e/SpotService/Shared/src/main/java/systems/zlink/e2e/spotservice/shared/UserActorLeaveHandler.java
package systems.zlink.e2e.spotservice.shared;

import java.util.concurrent.CompletionStage;
import systems.zlink.framework.handlers.ZLinkSpotActorSend;
import systems.zlink.framework.spots.ZLinkSpotActorSendContext;

public final class UserActorLeaveHandler {
    @ZLinkSpotActorSend(packetName = "LeaveActorReq")
    public CompletionStage<Void> handle(
        UserSpot spot,
        ScenarioActor actor,
        ZLinkSpotActorSendContext context,
        Contracts.LeaveActorReq request) {
        if (!request.actorId().equals(actor.actorId())) {
            throw new IllegalStateException("leave request actor does not match dispatched actor");
        }
        return spot.context().leaveActor(actor);
    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Server/Session/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/session/handlers/UserActorLeaveHandler.kt
package systems.zlink.e2e.kotlin.spotservice.session.handlers

import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.session.spots.ScenarioActor
import systems.zlink.e2e.kotlin.spotservice.session.spots.UserSpot
import systems.zlink.framework.CancellationToken
import systems.zlink.framework.handlers.ZLinkSpotActorRequest
import systems.zlink.framework.spots.ZLinkSpotActorRequestContext

class UserActorLeaveHandler {
    @ZLinkSpotActorRequest(packetName = "LeaveActorRequest")
    fun handle(
        spot: UserSpot,
        actor: ScenarioActor,
        context: ZLinkSpotActorRequestContext,
        request: Contracts.LeaveActorRequest,
        cancellationToken: CancellationToken
    ): Contracts.LeaveActorReply {
        if (request.actorId != actor.actorId()) {
            throw IllegalStateException("leave request actor does not match dispatched actor")
        }
        spot.context().leaveActor(actor).toCompletableFuture().join()
        return Contracts.LeaveActorReply(actor.actorId(), true)
    }
}
